### PR TITLE
 TRE API consumes and handles the StepResult event

### LINKED ISSUE #2013: Airlock - TRE API consumes and handles the StepResult event
As a TRE developer 
I want to add support in the TRE API for consuming StepResult event
So that whenever a new result arrives TRE API acts upon it

## Acceptance criteria

- [ ] TRE API should listen on the SB queue of step result
- [ ] TRE API updates cosmos with the newly requested status
- [ ] TRE API sends to Event grid a status changed event (logic is similar to what is done on submit request)

### LINKED ISSUE #2139: Airlock - approve reject - fix input string
**Describe the bug**

As we introduced new statuses such as ApprovalInProgress and RejectionInProgress.
The string we pass to the create review endpoint is 'approved' / 'rejected' which is not the correct stage.

since we dont want the user to pass 'approval_in_progress' / 'rejection_in...' we will replace this with 
approved: true/false

and in the api code, map it to the correct underlying status (which is then free to change when needed / if needed)

### PR DESCRIPTION
**Pull request by anatbal**
Fixes #2013 Fixes #2139

## What is being addressed

- Listener for stepResult event sb queue
- Changing the approval status

### CODE REVIEW

File: api_app/service_bus/airlock_request_status_update.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        app ([FastAPI]): Handle to the currently running app
+    """
+    receive_message_gen = receive_message_from_step_result_queue()
+
+    try:
+        async for message in receive_message_gen:
+            airlock_request_repo = AirlockRequestRepository(get_db_client(app))
+            result = await update_status_in_database(airlock_request_repo, message)
+            await receive_message_gen.asend(result)
+    except StopAsyncIteration:  # the async generator when finished signals end with this exception.
```
Comment (eladiw) [inline api_app/service_bus/airlock_request_status_update.py:97]: consider add logs all around.

File: api_app/service_bus/airlock_request_status_update.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                        logging.error(f"{strings.STEP_RESULT_MESSAGE_FORMAT_INCORRECT}: {e}")
+
+                    if result:
+                        logging.info(f"Received step_result status update message with correlation ID {msg.correlation_id}: {message}")
+                        await receiver.complete_message(msg)
+
+
+async def update_status_in_database(airlock_request_repo: AirlockRequestRepository, step_result_message: StepResultStatusUpdateMessage):
+    """
+    If there is another step in the operation after this one, process the substitutions + patch, then enqueue a message to process it.
```
Comment (eladiw) [inline api_app/service_bus/airlock_request_status_update.py:52]: what does this means

File: api_app/service_bus/airlock_request_status_update.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        new_status = AirlockRequestStatus(step_result_data.new_status)
+        # Find the airlock request by id
+        airlock_request = await get_airlock_request_by_id_from_path(airlock_request_id=airlock_request_id, airlock_request_repo=airlock_request_repo)
+        # Validate that the airlock request status is the same as current status
+        if airlock_request.status == current_status:
+            # update to new status and send to event grid
+            await update_status_and_publish_event_airlock_request(airlock_request=airlock_request, airlock_request_repo=airlock_request_repo, user=airlock_request.user, new_status=new_status)
+            result = True
+        else:
+            logging.error(strings.STEP_RESULT_MESSAGE_INVALID_STATUS)
```
Comment (eladiw) [inline api_app/service_bus/airlock_request_status_update.py:68]: you can add more details to the error. request id status, new status etc.

do you need to set the result here as well to true?

**Review by eladiw — CHANGES_REQUESTED** [review]
some comments

File: api_app/service_bus/airlock_request_status_update.py (hunk 3 above)
Comment (anatbal) [reply api_app/service_bus/airlock_request_status_update.py]: Changed

File: api_app/service_bus/airlock_request_status_update.py (hunk 2 above)
Comment (anatbal) [reply api_app/service_bus/airlock_request_status_update.py]: Left by mistake, changed

File: api_app/service_bus/airlock_request_status_update.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    Receives messages from the step result eventgrid topic and updates the status for
+    the airlock request in the state store.
+    Args:
+        app ([FastAPI]): Handle to the currently running app
+    """
+    receive_message_gen = receive_message_from_step_result_queue()
+
+    try:
+        async for message in receive_message_gen:
+            airlock_request_repo = AirlockRequestRepository(get_db_client(app))
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/service_bus/airlock_request_status_update.py:99]: I logged almost everything besides this line, which is simply getting the db client which should not matter for the user

File: api_app/service_bus/airlock_request_status_update.py (hunk 1 above)
Comment (anatbal) [reply api_app/service_bus/airlock_request_status_update.py:104]: Added

**Review by eladiw — APPROVED** [review]

**Pull request merged**